### PR TITLE
Refactor tink worker: abstraction around all Tink server calls

### PR DESCRIPTION
**Pull request by Raj-Dharwadkar**
Wrapper for tinker server calls that are made for tink worker to
process workflow and actions.

### CODE REVIEW

File: cmd/tink-worker/tink/tink.go [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+
+type Wdata struct {
+	Data WorkflowData
+}
+
+type WorkflowData struct {
+	WorkerID     string
+	WorkflowID   string
+	ActionStatus *pb.WorkflowActionStatus
+}
```
Comment (jacobweinstock) [inline cmd/tink-worker/tink/tink.go]: These fields aren't shared across any of the methods you have defined. They don't feel cohesive at an individual method level. meaning, As a caller of `GetWorkflowContexts`, I don't think I should need to understand if I need `ActionStatus` or not. You might think about making these field values parameters to their respective methods. thoughts?

File: cmd/tink-worker/tink/tink_test.go [hunk 2]
```diff
@@ -0,0 +1,264 @@
...
+	response     pb.WorkflowService_GetWorkflowContextsClient
+	actionstatus pb.WorkflowActionStatus
+}
+
+func (mcli *mockCli) GetWorkflowContexts(_ context.Context, _ *pb.WorkflowContextRequest, _ ...grpc.CallOption) (pb.WorkflowService_GetWorkflowContextsClient, error) {
+	return mcli.response, nil
+}
+
+func (mcli *mockCli) GetWorkflowActions(_ context.Context, _ *pb.WorkflowActionsRequest, _ ...grpc.CallOption) (*pb.WorkflowActionList, error) {
+	return nil, nil
```
Comment (jacobweinstock) [inline cmd/tink-worker/tink/tink_test.go]: might think about testing a non-nil error return. as a side effect, you could get 100% code cover by doing so.

File: cmd/tink-worker/tink/tink.go [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+
+// GetWorkflowContexts gets workflow context for worker id on success else returns error.
+func (w Wdata) GetWorkflowContexts(ctx context.Context, client pb.WorkflowServiceClient) (pb.WorkflowService_GetWorkflowContextsClient, error) {
```
Comment (jacobweinstock) [inline cmd/tink-worker/tink/tink.go]: might think about removing `Get`. 
> There's nothing wrong with providing getters and setters yourself, and it's often appropriate to do so, but it's neither idiomatic nor necessary to put Get into the getter's name. 

--- https://go.dev/doc/effective_go#Getters

File: cmd/tink-worker/tink/tink.go [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+	if w.Data.WorkerID == "" {
+		return nil, errors.New("Empty string is not a valid worker id")
+	}
+
+	if client == nil || (reflect.ValueOf(client).IsNil()) {
+		return nil, errors.New("WorkflowServiceClient interface is not valid")
```
Comment (jacobweinstock) [inline cmd/tink-worker/tink/tink.go]: might think about adding why the interface is not valid.

**Review by jacobweinstock — DISMISSED** [review]

File: cmd/tink-worker/tink/tink.go [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+	ActionStatus *pb.WorkflowActionStatus
+}
+
+// GetWorkflowContexts gets workflow context for worker id on success else returns error.
+func (w Wdata) GetWorkflowContexts(ctx context.Context, client pb.WorkflowServiceClient) (pb.WorkflowService_GetWorkflowContextsClient, error) {
+	if w.Data.WorkerID == "" {
+		return nil, errors.New("Empty string is not a valid worker id")
+	}
+
+	if client == nil || (reflect.ValueOf(client).IsNil()) {
```
Comment (stephen-fox) [inline cmd/tink-worker/tink/tink.go]: I think the `reflect.ValueOf(client).isNil()` is redundant here (and in other places it is made).

The `client == nil` condition is sufficient.

File: cmd/tink-worker/tink/tink.go [hunk 6]
```diff
@@ -0,0 +1,70 @@
...
+		return nil, errors.New("WorkflowServiceClient interface is not valid")
+	}
+
+	response, err := client.GetWorkflowContexts(ctx, &pb.WorkflowContextRequest{WorkerId: w.Data.WorkerID})
+	if err != nil {
+		return nil, errors.Wrap(err, "Failed to get workflow contexts")
```
Comment (stephen-fox) [inline cmd/tink-worker/tink/tink.go]: Maybe @jacobweinstock could weigh in on this - I think Go error messages should always start with a lower case letter. The contents of the error can contain capitalized letters as needed, but I think keeping as much of the error lower cased as possible helps with readability and searching during trouble shooting.

This blog post comes to mind, but it doesn't mention anything about capitalization: https://go.dev/blog/error-handling-and-go

**Review by stephen-fox — DISMISSED** [review]

File: cmd/tink-worker/tink/tink.go (hunk 6 above)
Comment (jacobweinstock) [reply cmd/tink-worker/tink/tink.go]: sure, I think that's something I've generally seen but I can't recall any one/place saying it's idiomatic. 


>When feasible, error strings should identify their origin, such as by having a prefix naming the operation or package that generated the error. For example, in package image, the string representation for a decoding error due to an unknown format is "image: unknown format".

--[ref](https://go.dev/doc/effective_go#:~:text=When%20feasible%2C%20error%20strings%20should%20identify%20their%20origin%2C%20such%20as%20by%20having%20a%20prefix%20naming%20the%20operation%20or%20package%20that%20generated%20the%20error.%20For%20example%2C%20in%20package%20image%2C%20the%20string%20representation%20for%20a%20decoding%20error%20due%20to%20an%20unknown%20format%20is%20%22image%3A%20unknown%20format%22.)

File: cmd/tink-worker/tink/tink.go (hunk 5 above)
Comment (Raj-Dharwadkar) [reply cmd/tink-worker/tink/tink.go]: As interface has type and value.  what if value is nil?

File: cmd/tink-worker/tink/tink.go [hunk 7]
```diff
@@ -0,0 +1,70 @@
...
+
+import (
+	"context"
+	"reflect"
+
+	"github.com/pkg/errors"
+	pb "github.com/tinkerbell/tink/protos/workflow"
+)
+
+type Wdata struct {
```
Comment (jacobweinstock) [inline cmd/tink-worker/tink/tink.go]: What's the purpose behind having this `Wdata` struct? would `WorkflowData` not satisfy on its own?

**Review by markjacksonfishing — APPROVED** [review]

File: cmd/tink-worker/tink/tink.go [hunk 8]
```diff
@@ -0,0 +1,61 @@
+package tink
```
Comment (ScottGarman) [inline cmd/tink-worker/tink/tink.go:1]: I think the target for this code should be adding it into `cmd/tink-worker/client/tink.go`, and moving the tests into `cmd/tink-worker/client/tink_tests.go`.

Comment (Raj-Dharwadkar) [reply cmd/tink-worker/tink/tink.go:1]: but As per this https://github.com/jacobweinstock/proposals/tree/0027-tink-worker-rearch/proposals/0027  this is in tink package ?

Comment (ScottGarman) [reply cmd/tink-worker/tink/tink.go:1]: I won't block the PR for this, but I think we're either going to need to move all of the code into one or the other of these locations. We can do that at a later time though.

**Review by ScottGarman — APPROVED** [review]

**Pull request merged**